File: trace/serve.py
#!/usr/bin/env python3
"""
trace/serve.py — the dev server, with caching turned off.

    python3 trace/serve.py            # http://localhost:8766
    python3 trace/serve.py 8790       # somewhere else

★ WHY THIS EXISTS INSTEAD OF `python3 -m http.server`.

Theodor, on a build where the panel CSS had already been written and shipped:
"when I press the arrow, nothing happens — everything just froze."

Nothing was broken. His browser was still running the PREVIOUS assets/app.css.
`http.server` sends `Last-Modified` and no `Cache-Control`, and a response with
no explicit freshness is one the browser is allowed to age itself — the
heuristic is a fraction of how long ago the file was modified, so a stylesheet
edited an hour ago can be treated as fresh for minutes without ever asking.

That produced the worst possible half-state, and it is worth understanding
because it is not obvious: index.html and the JS reloaded, the panel CSS did
not. So `openFrom()` ran, the route resolved, `is-locked` went on <html> and
<body> — and the page stopped scrolling, while the panel it had opened was
still an unstyled block sitting at the bottom of the document. A frozen page
with nothing visibly different on it. Exactly what he described, from a fix
that was already correct on disk.

`no-store` on every response makes that class of ghost impossible: what is on
disk is what runs, always, with no reasoning about revalidation. It costs
nothing locally — the files are on the same machine.

For production, `trace/bundle.py` stamps the asset URLs in the .dc.html files
with a content hash, which is the same guarantee by a different route.
"""
import functools, os, sys
from http.server import SimpleHTTPRequestHandler, ThreadingHTTPServer

ROOT = os.path.dirname(os.path.dirname(os.path.abspath(__file__)))
PORT = int(sys.argv[1]) if len(sys.argv) > 1 else 8766


class NoCacheHandler(SimpleHTTPRequestHandler):
    # HTTP/1.1 so keep-alive works; the suite opens a lot of pages at once
    protocol_version = "HTTP/1.1"

    def end_headers(self):
        self.send_header("Cache-Control", "no-store, no-cache, must-revalidate, max-age=0")
        self.send_header("Pragma", "no-cache")
        self.send_header("Expires", "0")
        super().end_headers()

    def log_message(self, fmt, *args):
        # one line per request is noise when a headless suite is hammering it;
        # errors still surface through the normal exception path
        pass


if __name__ == "__main__":
    handler = functools.partial(NoCacheHandler, directory=ROOT)
    with ThreadingHTTPServer(("", PORT), handler) as httpd:
        print(f"Field Atlas 2.0 — http://localhost:{PORT}/  (no-store; ^C to stop)")
        try:
            httpd.serve_forever()
        except KeyboardInterrupt:
            print("\nstopped")

File: trace/serve.mjs
/* ===========================================================================
 * trace/serve.mjs — the dev server, with caching turned off. Node, no deps.
 *
 *     node trace/serve.mjs            # http://localhost:8766
 *     node trace/serve.mjs 8790       # somewhere else
 *
 * ★ WHY THIS EXISTS ALONGSIDE serve.py.
 *
 * `trace/serve.py` is the original and the two are interchangeable — same root,
 * same port, same headers. It is here because Python is not installed on every
 * machine this repo gets opened on, and a project whose whole selling point is
 * "no build step, just serve it" should not be un-runnable for want of an
 * interpreter it never actually needs. Node is already required by
 * trace/verify.mjs and trace/shots.mjs.
 *
 * Read the long note at the top of serve.py for why `no-store` is not
 * optional here: a half-reloaded page — new JS against old CSS — produced a
 * frozen, unstyled panel that looked exactly like a bug in code that was
 * already correct on disk.
 * ======================================================================== */
import { createServer } from 'node:http';
import { createReadStream, statSync } from 'node:fs';
import { join, normalize, extname, dirname } from 'node:path';
import { fileURLToPath } from 'node:url';

const ROOT = dirname(dirname(fileURLToPath(import.meta.url)));
const PORT = Number(process.argv[2]) || 8766;

/* Only the types this site actually serves. An unknown extension gets
   octet-stream rather than a guess: a mislabelled ES module is refused by the
   browser outright, which is a far more confusing failure than a download. */
const TYPES = {
  '.html': 'text/html; charset=utf-8',
  '.js': 'text/javascript; charset=utf-8',
  '.mjs': 'text/javascript; charset=utf-8',
  '.css': 'text/css; charset=utf-8',
  '.json': 'application/json; charset=utf-8',
  '.webmanifest': 'application/manifest+json; charset=utf-8',
  '.svg': 'image/svg+xml',
  '.png': 'image/png',
  '.jpg': 'image/jpeg',
  '.jpeg': 'image/jpeg',
  '.webp': 'image/webp',
  '.woff2': 'font/woff2',
  '.ico': 'image/x-icon',
  '.txt': 'text/plain; charset=utf-8',
  '.md': 'text/plain; charset=utf-8',
};

const NO_CACHE = {
  'Cache-Control': 'no-store, no-cache, must-revalidate, max-age=0',
  Pragma: 'no-cache',
  Expires: '0',
};

createServer((req, res) => {
  let path;
  try {
    path = decodeURIComponent(new URL(req.url, 'http://x').pathname);
  } catch {
    res.writeHead(400, NO_CACHE); res.end('bad request'); return;
  }

  /* Resolve inside ROOT and check that it stayed there. `..` in a URL is the
     one thing a static server must not honour, and normalize() alone does not
     stop it — the containment test does. */
  let file = normalize(join(ROOT, path));
  if (!file.startsWith(ROOT)) { res.writeHead(403, NO_CACHE); res.end('forbidden'); return; }

  let st;
  try { st = statSync(file); } catch { st = null; }
  if (st?.isDirectory()) {
    file = join(file, 'index.html');
    try { st = statSync(file); } catch { st = null; }
  }
  if (!st?.isFile()) { res.writeHead(404, NO_CACHE); res.end('not found'); return; }

  res.writeHead(200, {
    ...NO_CACHE,
    'Content-Type': TYPES[extname(file).toLowerCase()] || 'application/octet-stream',
    'Content-Length': st.size,
  });
  if (req.method === 'HEAD') { res.end(); return; }
  createReadStream(file).pipe(res);
}).listen(PORT, () => {
  console.log(`Field Atlas 2.0 — http://localhost:${PORT}/  (no-store; ^C to stop)`);
});
